### PR TITLE
Pull Request (crudSprint3 to integration)

### PR DESCRIPTION
**Pull request by xaterz**
What has been implemented in crudSprint3:
* Show list of modified modules, and what has been modified
* Select the type of modified modules to display based on the type of modified information (mounting, quota, name, description or MC)
* Restoration of module information
* Fixed bug where new modules appear in fixed mounting
* Added mounting icons in module overview and individual mounting view
* Display of outcome messages are no longer handled by sessions, but by a new class called Outcome

NOTE: Please delete and repopulate database before testing as schema has been modfied.

### CODE REVIEW

File: components/handlers/fixed_module_mountings.py [hunk 1]
```diff
@@ -18,22 +18,13 @@ class Fixed(object):
     def __init__(self):
         '''
             Full_mounting_plan is a list of 'subplans'
-            Each subplan is a list of 4 attributes (code, name, sem 1 mounting, sem 2 mounting)
+            Each subplan is a list of 5 attributes (code, name, sem 1 mounting, sem 2 mounting, status)
```
Comment (tgqiang) [inline components/handlers/fixed_module_mountings.py]: Pylint: Line too long

File: components/handlers/fixed_module_mountings.py [hunk 2]
```diff
@@ -86,9 +79,13 @@ def GET(self):
 
         self.populate_module_code_and_name()
         self.populate_mounting_values()
-        current_ay = self.get_current_ay()
+        current_ay = model.get_current_ay()
+
+        full_mounting_plan = self.full_mounting_plan
+        full_mounting_plan = [subplan for subplan in full_mounting_plan
```
Comment (tgqiang) [inline components/handlers/fixed_module_mountings.py]: Pylint: Trailing whitespace at line 85, line too long at line 86

File: components/handlers/index.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 from app import RENDER, SESSION
 import web
 from components import model
+from components.handlers.outcome import Outcome
```
Comment (tgqiang) [inline components/handlers/index.py]: Pylint: Unused import statements, please remove

File: components/handlers/modified_modules.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        for module in modules_wth_modified_mountings:
+            mounting_change = module[3]
+            if mounting_change == 1:   # Unmounted --> Mounted
+                code = module[0]
+                current_ay = module[1]
+                target_ay = module[2]
+                quota = model.get_quota_of_target_tenta_ay_sem(code, target_ay)
+                if quota is not None and quota > 0:
+                    modified_modules.append((code, current_ay, target_ay,
+                                            "Unmounted", quota, '+'+str(quota)))
```
Comment (tgqiang) [inline components/handlers/modified_modules.py]: Pylint:
1. Wrong continued indentation at lines 125, 133
2. Trailing whitespace at lines 230, 243, 244

Please fix

File: components/handlers/module_edit.py [hunk 5]
```diff
@@ -79,10 +90,5 @@ def POST(self):
             elif mounting_status == "Not Mounted":
                 outcome = model.delete_tenta_mounting(module_code, ay_sem)
 
-            if outcome is True:
-                SESSION['editMountMsg'] = "Mounting info edited sucessfully!"
-            else:
-                SESSION['editMountMsg'] = "Sorry, an error has occurred!"
+            return Outcome().POST("edit_mounting", outcome, module_code, ay_sem)
```
Comment (tgqiang) [inline components/handlers/module_edit.py]: Pylint: trailing newlines

Please fix.

File: components/handlers/module_listing.py [hunk 6]
```diff
@@ -54,19 +48,24 @@ def POST(self):
...
+                return RENDER.moduleListing(modules, form)
 
-        # else add module to db and refresh page
-        outcome = model.add_module(form.d.code, form.d.name, form.d.description, form.d.mc, 'New')
-        if outcome is False:
-            SESSION['keyError'] = True
-            SESSION['displayErrorMessage'] = True
-        raise web.seeother(self.URL_THIS_PAGE)        # load index.html again
+            # else add module to db and refresh page
+            outcome = model.add_module(form.d.code, form.d.name, form.d.description, form.d.mc, 'New')
```
Comment (tgqiang) [inline components/handlers/module_listing.py]: Pylint: line too long

File: components/handlers/module_overview.py [hunk 7]
```diff
@@ -86,12 +99,14 @@ def load_tenta_mounting_plan(self, module_code):
 
         self.tenta_mounting_plan = tenta_mounting_plan
 
+
     def get_overlapping_mods(self,code):
```
Comment (tgqiang) [inline components/handlers/module_overview.py]: Pylint: 1 whitespace needed after <code>(self,</code>

File: components/handlers/module_restore.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
+'''
```
Comment (tgqiang) [inline components/handlers/module_restore.py]: Pylint: please do linting and fix the corresponding linting issues for this.

File: components/handlers/outcome.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+'''
+    This module contains the handler for displaying the outcome of a form submission
+    (E.g. adding/editing/restoring a module)
+'''
+
+from app import RENDER, SESSION
+import web
```
Comment (tgqiang) [inline components/handlers/outcome.py]: Unused imports <code>web, model</code>

File: components/handlers/outcome.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+            outcome_message = None
+            redirect_page = None
+
+            if action == "add_module":
+                if outcome is True:
+                    outcome_message = "Module " + module_code + " has been added successfully!"
+                    redirect_page = "/viewModule?code="+module_code
+                else:
+                    outcome_message = "Error: Module code already exists! Please use another module code."
```
Comment (tgqiang) [inline components/handlers/outcome.py]: Pylint: Line too long.

Please fix

File: components/handlers/outcome.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+import web
+from components import model
+
+
+class Outcome(object):
+    '''
+        This class handles the display of a form submission's outcome
+    '''
+    def POST(self, *received_data):
+        if received_data is not None:
```
Comment (tgqiang) [inline components/handlers/outcome.py]: Pylint: Missing docstring.

File: components/model.py [hunk 12]
```diff
@@ -116,24 +118,60 @@ def get_tenta_mounting_and_quota(code):
...
 
+def get_mounting_of_target_fixed_ay_sem(code, ay_sem):
+    '''
+        Get the mounting status of a module in a target fixed AY/Sem
+    '''
+    sql_command = "SELECT COUNT(*) FROM moduleMounted " +\
+                  "WHERE moduleCode=%s AND acadYearAndSem=%s"
+    DB_CURSOR.execute(sql_command, (code, ay_sem))
+    result = DB_CURSOR.fetchone()
+    if result[0] == 1:
```
Comment (tgqiang) [inline components/model.py]: Replace this with return result[0] == 1

File: test/test_edit_module_fixture.py [hunk 13]
```diff
@@ -65,13 +65,11 @@ def test_module_edit_submit_response(self):
         edit_submit_button = goto_edit_response.forms__get()[self.EDIT_MODULE_FORM_NAME]
```
Comment (nlzz22) [inline test/test_edit_module_fixture.py]: Trailing newline at line 100

**Comment by tgqiang** [comment]
image

Module description was **not changed** but it is flagged as **changed**.

This requires bug-fix.

**Comment by nlzz22** [comment]
please pylint test/test_module_modified_backend.py

**Comment by tgqiang** [comment]
Bug-fix is implemented and linting issues resolved, pull request approved.

**Pull request merged**